[Comments section] Update the comment list

juancoob opened the pull request:
## Fixes #15397

An automattician reported the comment list did not update after comment editing. Opening the comment details from the list and editing the comment section didn't reflect the changes on the main list, so it is needed a PTR on the list to make the changes visible.

This PR is going to update the main comment list when the user comes back from the comment detail screen.

As suggested @khaykov in the comment section, I'm going to:

- [x] Leave `localCommentCacheUpdateHandler` as a constructor parameter
- [x] Remove the delay, use only `performSingleCommentModeration` and understand better its code
- [x] Change the `SharedPreference` strategy
- [x] Add unit testing to test the new changes

## To test:

- Go in My Site > Comments in a site with at least one comment
- Tap on a comment to open the comment details view
- From the overflow menu select Edit, make some changes and save them
- Notice you are landed in comment details and that is updated
- Go back to the comments list and notice the original comment entry in the list is correctly updated

## Regression Notes
1. Potential unintended areas of impact:
When the comment is moderated and edited at the same time, the main list could not refresh the updated comment.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
I have updated the test `onActionMenuClicked triggers DONE action if comment update successfully` to verify if the `requestCommentsUpdate` method is called from `localCommentCacheUpdateHandler`

## PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

juancoob committed:
Update CommentsDetailsActivity to refresh the comment list

CommentsDetailsActivity has been changed to update the main comment list when the user comes back. It could be by using the up functionality or the back button.

juancoob committed:
Remove repeated version

lifecycleVersion was removed because it was defined some lines above

juancoob committed:
Revert CommentsDetailActivity

Revert the file to find out a better way to update the comment list

juancoob committed:
Update CommentDetailFragment to store comment data

Uses SharedPreferences to manage the states from the comments: When it's edited and moderated

juancoob committed:
Update CommentsDetailActivity to manage the comment refresh on the list

When the user comes back to the main comment list, the code checks if the comment has been updated. In that case, pass the commentId and the status to the intent.

juancoob committed:
Update UnifiedCommentListFragment to manage registerForActivityResult

When registerForActivityResult retrieves the response, the code manages the comment edit state and the comment moderation state to refresh the comment list.
The delay is needed to avoid a race condition whether both updates are executed.

khaykov commented on build.gradle:
```diff
@@ -140,7 +140,6 @@ ext {
     preferenceVersion = '1.1.0'
     swipeToRefresh = '1.1.0'
     uCropVersion = '2.2.4'
-    lifecycleVersion = '2.2.0'
```
Good catch 👍

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/comments/unified/UnifiedCommentListFragment.kt:
```diff
@@ -167,7 +173,27 @@ class UnifiedCommentListFragment : Fragment(R.layout.unified_comment_list_fragme
+                        COMMENT_STATE_INITIAL
+                    ) == COMMENT_STATE_SUCCESSFULLY_EDITED
+                ) {
+                    viewModel.localCommentCacheUpdateHandler.requestCommentsUpdate()
```
I don't think this is the best way to do it, to be honest. No one needs to know that VM is using `localCommentCacheUpdateHandler`. If you want the logic to be related to VM, I would expose a public method from the VM, that will do the update. This way we can at least test it.

If you think that the logic is not related to VM, I would have just used a different instance `localCommentCacheUpdateHandler` not exposed from VM.

juancoob replied:
I'll check that, thanks 😄

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/comments/unified/UnifiedCommentListFragment.kt:
```diff
@@ -167,7 +173,27 @@ class UnifiedCommentListFragment : Fragment(R.layout.unified_comment_list_fragme
+
+                    // If the user does a comment moderation next, add a short delay to avoid a
+                    // race condition updating data on the main comment list.
+                    delay(10)
```
Delay is a bit of a hacky way to address a race condition, in my opinion. I would try to avoid it, to begin with. Also, looking at the code, it seems that if the comment is moderated and edited at the same time, the `performSingleCommentModeration` should have taken care of updating the list from local cache by calling `requestCommentsUpdate` before network request.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/comments/unified/UnifiedCommentListFragment.kt:
```diff
@@ -167,7 +173,27 @@ class UnifiedCommentListFragment : Fragment(R.layout.unified_comment_list_fragme
         if (response != null) {
-            viewModel.performSingleCommentModeration(response.commentId, response.commentStatus)
+            viewLifecycleOwner.lifecycleScope.launchWhenStarted {
+                if (sharedPreferences?.getString(
```
Since we are performing multiple actions here, personally I would bring all this logic into the VM, so we can test it with injected preferences provider.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -138,6 +139,16 @@
+    public static final String COMMENT_FILE_KEY = "commentStateFileKey";
+
+    public static final String COMMENT_STATE_KEY = "commentStateKey";
+    public static final String COMMENT_STATE_INITIAL = "commentStateInitial";
```
If you expect more statuses in the future, I think it would make sense to have them as enum or similar. If not, do you think boolean would work here?

juancoob replied:
Hey @khaykov I prefer an enum, so I'll update it 😄

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/comments/unified/UnifiedCommentListViewModel.kt:
```diff
@@ -69,7 +69,7 @@ class UnifiedCommentListViewModel @Inject constructor(
     @Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher,
     private val unifiedCommentsListHandler: UnifiedCommentsListHandler,
-    localCommentCacheUpdateHandler: LocalCommentCacheUpdateHandler
+    val localCommentCacheUpdateHandler: LocalCommentCacheUpdateHandler
```
As I mentioned before, we should not expose this.

khaykov reviewed (changes requested):
Hey @juancoob 👋 Thank you for the PR! 

While your approach with using `SharedPreferences` for status management and fragment communication works in practice, I think there are better ways to do this.

Regarding testing, even with your approach there are multiple opportunities for automated testing.

Additionally, I would love to see more details about the direction you took in the PR description.

juancoob committed:
Revert previous changes to create use a simpler approach

juancoob committed:
Update UnifiedCommentsEditViewModel

When the comment is edited successfully, the app refreshes it.

juancoob committed:
Update UnifiedCommentsEditViewModelTest to change tests

The VM instance needed the "LocalCommentCacheUpdateHandler" mock instance,  and the test "onActionMenuClicked triggers DONE action if comment update successfully", was changed to verify if the method "requestCommentUpdate" was called from the class "LocalCommentCacheUpdateHandler"

AliSoftware commented:
FYI: Moving this PR to the 19.1 milestone, as I'm doing the code-freeze for 19.0 today.

khaykov commented:
Going to close this one for now.
